google-cloud-spanner/src/test/java/com/google/cloud/spanner/MockSpannerServiceImpl.java

```diff
@@ -1120,10 +1125,20 @@ public void executeBatchDml(
       }
       ExecuteBatchDmlResponse.Builder builder = ExecuteBatchDmlResponse.newBuilder();
       for (StatementResult res : results) {
+        Long updateCount;
+        switch (res.getType()) {
+          case UPDATE_COUNT:
+            updateCount = res.getUpdateCount();
+            break;
+          case RESULT_SET:
+            updateCount = res.getResultSet().getStats().getRowCountExact();
+            break;
+          default:
+            throw new IllegalStateException("Invalid result type: " + res.getType());
+        }
```

Comment:
Note that while this works fine here and is semantically correct, it might be a little confusing, because (I think) the backend does not allow DML statements with a RETURNING clause to be executed using `BatchDml`.

Reply:
DML with Returning clause over BatchDML is allowed. Have added corresponding integration tests for `executeBatchDml` and `executeBatchDmlAsync` in ITDmlReturningTest in this PR.